### PR TITLE
feat: Add version information panel (#4312)

### PR DESCRIPTION
**Pull request by tetchel**
- Add new extended version info panel that pops up when the version is clicked in the navbar
- Refactor kustomize and helm versions to be more user-friendly (also affects argo CLI)

Signed-off-by: Tim Etchells <tetchell@redhat.com>

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates? 
 - Maybe
* [ ] I've updated documentation as required by this PR.
* [x] Optional. My organization is added to USERS.md.
* [x] I've signed the CLA and my build is green ([troubleshooting builds](https://argoproj.github.io/argo-cd/developer-guide/ci/)).

### CODE REVIEW

**Comment by tetchel** [comment]
new `argocd version` server output

```
argocd-server: v99.99.99+unknown
  BuildDate: 1970-01-01T00:00:00Z
  GitCommit:
  GitTreeState:
  GoVersion: go1.14.1
  Compiler: gc
  Platform: linux/amd64
  Ksonnet Version: v0.13.1
  Kustomize Version: v3.8.1 2020-07-16T00:58:46Z
  Helm Version: v3.3.1+g249e521
  Kubectl Version: v1.17.8
```

**Comment by tetchel** [comment]
demo:
version-panel(1)

[version-panel.mov.zip](https://github.com/argoproj/argo-cd/files/5247852/version-panel.mov.zip)

**Comment by tetchel** [comment]
resolves https://github.com/argoproj/argo-cd/issues/4312

**Comment by tetchel** [comment]
I am totally new to Go and fairly new to React so please go ahead and nitpick if you'd like :)

File: ui/src/app/version-info/components/version-info-panel.tsx [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    }
+
+    public render() {
+        return (
+            <DataLoader load={() => this.props.version}>
+                {version => {
+                    return (
+                        <SlidingPanel header={this.header} isShown={this.props.isShown} onClose={() => this.props.onClose()} hasCloseButton={true} isNarrow={true}>
+                            <div className='version-info-table argo-table-list'>
+                                {/* <div className='argo-table-list__head'>
```
Comment (rbreeze) [inline ui/src/app/version-info/components/version-info-panel.tsx:29]: Can you remove this comment?

File: ui/src/app/version-info/components/version-info.scss [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    white-space: nowrap;
+    cursor: pointer;
+
+    &:hover, &:focus {
+        text-decoration: underline;
+    }
+}
+
+.version-copy-btn-container {
+    margin-top: 15px;   // half of the panel's padding
```
Comment (rbreeze) [inline ui/src/app/version-info/components/version-info.scss:13]:
```suggestion
    margin-top: 15px;
```

File: ui/src/app/version-info/components/version-info-panel.tsx [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+     * Formats the version data and renders the table rows.
+     */
+    private buildVersionTable(version: VersionMessage): JSX.Element {
+        // match the order/format of `argocd version`
+        // but leave out 'version' from the titles; that's implied by us being in the version info panel.
+        // These key/values are rendered to the user as written in this object
+        const formattedVersion = {
+            'Argo CD': version.Version,
+            'Build Date': version.BuildDate,
+            // 'Git Commit':
```
Comment (rbreeze) [inline ui/src/app/version-info/components/version-info-panel.tsx:64]:
```suggestion          
```

File: ui/src/app/version-info/components/version-info-panel.tsx [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+            // 'Git Tree State':
```
Comment (rbreeze) [inline ui/src/app/version-info/components/version-info-panel.tsx:65]:
```suggestion
```

File: ui/src/app/version-info/components/version-info-panel.tsx [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+                                    <div className='row'>
+                                        <div className='columns small-4'>Tool</div>
+                                        <div className='columns small-4'>Version</div>
+                                    </div>
+                                </div> */}
+                                {this.buildVersionTable(version)}
+                            </div>
+                            <div className='version-copy-btn-container'>
+                                <Tooltip content='Copy all version info as JSON'>
+                                    <CopyToClipboard text={JSON.stringify(version, undefined, 4)} onCopy={() => this.onCopy()}>
```
Comment (rbreeze) [inline ui/src/app/version-info/components/version-info-panel.tsx:39]: Could we leave out the `copy-to-clipboard` library? 

```{javascript}
onClick={() => {navigator.clipboard.writeText(TEXT_TO_COPY)}}
```

Most browsers support this now: https://caniuse.com/?search=clipboard
I don't think it's worth it to include a 37kB library for something this small.

File: ui/src/app/version-info/components/version-info.scss [hunk 6]
```diff
@@ -0,0 +1,24 @@
+@import 'node_modules/argo-ui/src/styles/config';
```
Comment (rbreeze) [inline ui/src/app/version-info/components/version-info.scss:1]:
```suggestion
```

**Review by rbreeze — CHANGES_REQUESTED** [review]
Remove the `copy-to-clipboard` library, and it LGTM.

File: ui/src/app/version-info/components/version-info-panel.tsx (hunk 5 above)
Comment (tetchel) [reply ui/src/app/version-info/components/version-info-panel.tsx:39]: yes I will use that API. I recall when I was looking at clipboard options none of the ones I saw had great support so I guess I missed this one at 96%.

**Comment by rbreeze** [comment]
Love the new panel. I found that after removing the CSS it still looks almost identical so I think it's best to leave it out if possible. I also inlined `VersionInfoButton` to keep down the amount of new code added, and I think the `VersionInfoPanel` file should live in the `shared` directory. 

LGTM, this is really useful.

**Comment by tetchel** [comment]
I can appreciate the desire to keep the code as simple as possible, but I wouldn't have added the styles if I didn't think those details mattered. 

Especially the styles on the `version-info-button` are important to allow a user to discover this panel; otherwise on mouse-over it looks like the tooltip is all you get. The `:hover` `pointer` and `underline` are what make it look clickable. 

The styles on the panel itself are definitely more subjective and I am OK without most of those, however I think the Copy button is a little cramped now and the icon could use some padding. 

image

The reason I set the `min-width` was because the text changes on the button when you click it, and I did not like the button changing widths on-click and then resetting.

**Comment by rbreeze** [comment]
I agree with everything you pointed out. I added back the styles to keep the copy button a consistent width, but inline because I don't think it warrants an additional file. I also added an additional `&nbsp;` to give it some extra space (the equivalent of `margin-right: 2ch`). 

For the version button itself, I don't think there's a good way to add `text-decoration: underline` on hover without making the version button a new component, but I really don't think this should be its own component because it is so small. By changing it to `<a>` rather than `<span>` we get the `cursor: pointer` for free (it follows accessibility guidelines this way too), and I think this combined with the tooltip is enough for users to discover it. I also looked around elsewhere in the UI to see if we could "steal" a style that underlines on hover, but there isn't anywhere that uses that design pattern so I additionally think for consistency's sake, the underline is unnecessary.

If you agree, I think it's ready to merge. Thanks for working on this!

**Comment by tetchel** [comment]
Personally I prefer new files for separate components even if they are simple, and [s]css files over inline styles and nbsp. 

But, that is subjective enough that I am happy to merge this.

**Comment by rbreeze** [comment]
I often agree, but in my experience with Argo the standard seems to be a preference toward minimal amount of files/components. Glad you're okay with the changes. Thanks!

**Review by rbreeze — APPROVED** [review]

**Pull request merged**